### PR TITLE
[jsx-no-script-url]: use shared settings to specify link components

### LINKED ISSUE #2865: Make jsx-no-script-url use shared settings to specify link components
Why do we have to specify Link components in `react/jsx-no-script-url` settings when we already have that in shared config? Can't shared config be used here, as it is in `react/jsx-no-target-blank`?

### PR DESCRIPTION
**Pull request by burtek**
Fixes #2865

### CODE REVIEW

**Comment by burtek** [comment]
I'm probably gonna need to change `getLinkComponents` util to allow multiple props for single component. You can't really give it `MyComponent` with two different props at this time, while `jsx-no-script-url` allows that. Am thinking about changing a string prop `linkAttribute` to a string array prop `linkAttributes` (and internally mapping `linkAttribute` to `linkAttributes: [linkAttribute]` for backwards compatibility).

File: docs/rules/jsx-no-script-url.md [hunk 1]
```diff
@@ -44,6 +46,11 @@ Examples of **correct** code for this rule:
 ```
 
 Allows you to indicate a specific list of properties used by a custom component to be checked.
+This will override anything passed to `linkComponents` setting.
+
+NOTE: This rule now takes into account `linkComponents` setting and it should be used as primary source of link components.
+The rule still allows passing link components as rule option, but it is meant only as backwards-compatibility feature.
+New setups should only use `linkComponents` setting.
```
Comment (ljharb) [inline docs/rules/jsx-no-script-url.md:53]:
```suggestion
NOTE: This rule now takes into account the `linkComponents` setting, which should be used as the sole source of truth for link components.
The rule still allows passing link components as rule option for backwards compatibility, but this option is deprecated.
```

File: lib/rules/jsx-no-script-url.js [hunk 2]
```diff
@@ -21,26 +22,23 @@ function hasJavaScriptProtocol(attr) {
...
 
-  const props = el.props || [];
-  return node.name && props.indexOf(name) !== -1;
+function parseLegacyOption(option) {
+  const config = linkComponentsUtil.getLinkComponents({}) // get defaults
+  option.forEach(function(opt) {
+    opt.props.forEach(function(prop) { // FIXME: only last prop will work at the moment
+      config.set(opt.name, prop)
+    })
+  })
```
Comment (ljharb) [inline lib/rules/jsx-no-script-url.js:40]:
```suggestion
  const config = linkComponentsUtil.getLinkComponents({}); // get defaults
  option.forEach((opt) => {
    opt.props.forEach((prop) => { // FIXME: only last prop will work at the moment
      config.set(opt.name, prop);
    });
  });
```

File: lib/rules/jsx-no-script-url.js [hunk 3]
```diff
@@ -77,16 +75,17 @@ module.exports = {
         },
         required: ['name', 'props'],
         additionalProperties: false,
+        deprecated: true, // ?
```
Comment (ljharb) [inline lib/rules/jsx-no-script-url.js:78]: sounds good to me. does eslint actually check this?

File: lib/rules/jsx-no-script-url.js [hunk 4]
```diff
@@ -77,16 +75,17 @@ module.exports = {
...
       },
     }],
   },
 
   create(context) {
-    const config = context.options[0] || [];
+    const linkComponents = context.options[0] ? parseLegacyOption(context.options[0]) : linkComponentsUtil.getLinkComponents(context);
```
Comment (ljharb) [inline lib/rules/jsx-no-script-url.js:84]: i think that perhaps this should just always start with the setting, and concatenate the option when present?

File: lib/rules/jsx-no-script-url.js (hunk 3 above)
Comment (burtek) [reply lib/rules/jsx-no-script-url.js:78]: [yes](https://eslint.org/docs/latest/extend/custom-rules#rule-structure)

File: README.md [hunk 5]
```diff
@@ -63,12 +63,14 @@ You should also specify settings that will be shared across all the plugin rules
     "formComponents": [
       // Components used as alternatives to <form> for forms, eg. <Form endpoint={ url } />
       "CustomForm",
-      {"name": "Form", "formAttribute": "endpoint"}
+      {"name": "Form", "formAttributes": ["registerEnpoint", "loginEnpoint"]}, // allows specifying multiple properties if necessary
```
Comment (ljharb) [inline README.md:66]:
```suggestion
      {"name": "Form", "formAttributes": ["registerEndpoint", "loginEndpoint"]}, // allows specifying multiple properties if necessary
```

File: README.md [hunk 6]
```diff
@@ -63,12 +63,14 @@ You should also specify settings that will be shared across all the plugin rules
...
+      {"name": "SimpleForm", "formAttribute": "endpoint"}, // deprecated backwards-compatible version
```
Comment (ljharb) [inline README.md:67]: i'm not sure we need a new option name here - can we just overload `formAttribute` to be either a string, or an array of strings?

Comment (burtek) [reply README.md:67]: We can actually. Not a bad idea.

File: lib/rules/jsx-no-script-url.js [hunk 7]
```diff
@@ -21,26 +22,20 @@ function hasJavaScriptProtocol(attr) {
...
-    return true;
-  }
+  if (!name || !parentName || !config.has(parentName)) return false;
 
-  const el = config.find((i) => i.name === parentName);
-  if (!el) {
-    return false;
-  }
+  const attributes = config.get(parentName)
+  return attributes.includes(name);
```
Comment (ljharb) [inline lib/rules/jsx-no-script-url.js:32]:
```suggestion
  return includes(attributes, name);
```

File: lib/rules/jsx-no-target-blank.js [hunk 8]
```diff
@@ -48,16 +48,16 @@ function attributeValuePossiblyBlank(attribute) {
...
   const foundExternalLink = linkIndex !== -1 && ((attr) => attr.value && attr.value.type === 'Literal' && /^(?:\w+:|\/\/)/.test(attr.value.value))(
     node.attributes[linkIndex]);
   return foundExternalLink || (warnOnSpreadAttributes && linkIndex < spreadAttributeIndex);
 }
 
-function hasDynamicLink(node, linkAttribute) {
+function hasDynamicLink(node, linkAttributes) {
   const dynamicLinkIndex = findLastIndex(node.attributes, (attr) => attr.name
-    && attr.name.name === linkAttribute
+    && linkAttributes.includes(attr.name.name)
```
Comment (ljharb) [inline lib/rules/jsx-no-target-blank.js:60]:
```suggestion
    && includes(linkAttributes, attr.name.name)
```

File: lib/rules/jsx-no-target-blank.js [hunk 9]
```diff
@@ -48,16 +48,16 @@ function attributeValuePossiblyBlank(attribute) {
   return false;
 }
 
-function hasExternalLink(node, linkAttribute, warnOnSpreadAttributes, spreadAttributeIndex) {
-  const linkIndex = findLastIndex(node.attributes, (attr) => attr.name && attr.name.name === linkAttribute);
+function hasExternalLink(node, linkAttributes, warnOnSpreadAttributes, spreadAttributeIndex) {
+  const linkIndex = findLastIndex(node.attributes, (attr) => attr.name && linkAttributes.includes(attr.name.name));
```
Comment (ljharb) [inline lib/rules/jsx-no-target-blank.js:52]:
```suggestion
  const linkIndex = findLastIndex(node.attributes, (attr) => attr.name && includes(linkAttributes, attr.name.name));
```

File: lib/util/linkComponents.js [hunk 10]
```diff
@@ -37,9 +37,9 @@ function getLinkComponents(context) {
   );
   return new Map(map(iterFrom(linkComponents), (value) => {
     if (typeof value === 'string') {
-      return [value, DEFAULT_LINK_ATTRIBUTE];
+      return [value, [DEFAULT_LINK_ATTRIBUTE]];
     }
-    return [value.name, value.linkAttribute];
+    return [value.name, value.linkAttributes || [value.linkAttribute]];
```
Comment (ljharb) [inline lib/util/linkComponents.js:42]:
```suggestion
    return [value.name, [].concat(value.linkAttribute)];
```

File: lib/util/linkComponents.js [hunk 11]
```diff
@@ -24,9 +24,9 @@ function getFormComponents(context) {
   );
   return new Map(map(iterFrom(formComponents), (value) => {
     if (typeof value === 'string') {
-      return [value, DEFAULT_FORM_ATTRIBUTE];
+      return [value, [DEFAULT_FORM_ATTRIBUTE]];
     }
-    return [value.name, value.formAttribute];
+    return [value.name, value.formAttributes || [value.formAttribute]];
```
Comment (ljharb) [inline lib/util/linkComponents.js:29]:
```suggestion
    return [value.name, [].concat(value.formAttribute)];
```

File: lib/rules/jsx-no-target-blank.js [hunk 12]
```diff
@@ -265,11 +266,11 @@ module.exports = {
...
           }
 
-          const formAttribute = formComponents.get(node.name.name);
+          const formAttributes = formComponents.get(node.name.name);
 
           if (
-            hasExternalLink(node, formAttribute)
-            || (enforceDynamicLinks === 'always' && hasDynamicLink(node, formAttribute))
+            hasExternalLink(node, formAttributes)
+            || (enforceDynamicLinks === 'always' && hasDynamicLink(node, formAttributes))
```
Comment (ljharb) [inline lib/rules/jsx-no-target-blank.js:273]: seems like we need code coverage for this line; any chance you could add it?

Comment (burtek) [reply lib/rules/jsx-no-target-blank.js:273]: Am planning to but I'm confused as I kinda think the test case for it is already there... Will take another look later.

Comment (burtek) [reply lib/rules/jsx-no-target-blank.js:273]: @ljharb should be good now

**Review by ljharb — APPROVED** [review]
This looks good.

I've split the PR into two commits - the first one is objectively semver-minor, and has the "multiple properties" feature.

The second, though, I'm a bit uncertain if it's semver-minor or semver-major. Could there be someone relying on `linkAttributes` *not* applying to this rule?

**Comment by burtek** [comment]
I don't know. Myself I have this plugin configured in multiple projects (both my own and at work) and we always had to duplicate the setup for this rule.

That being said, that's why the first implementation had the rule option override global settings - that way it wouldn't be a breaking change.

**Comment by ljharb** [comment]
hmm.

even for the case where the rule option wasn't passed, though, someone might have the link components setting and expect the rule NOT to warn on them, no?

**Comment by burtek** [comment]
Yeah, you might be right actually.

I find with this PR being marked as semver-major

**Comment by ljharb** [comment]
Unfortunately that'd delay it from landing probably indefinitely.

How about instead, we modify the second commit so that an *additional* option is required, a boolean which defaults to `false`, that enables hooking in the linkComponents (whether the existing option is provided or not)?

**Comment by burtek** [comment]
I can do that

**Comment by burtek** [comment]
@ljharb how about now? :)

**Review by ljharb — APPROVED** [review]

**Pull request merged**